### PR TITLE
ComposeScene.roots to return child compose scenes created for Window and DialogWindow

### PR DESCRIPTION
**Pull request by m-sasha**
Currently, the `ComposeScene` created for a `Window` or a `DialogWindow` can't be reached by semantic node fetching, because we don't return their owners from `testOwner.getRoots` (in `ComposeScene.roots`). So for example, this will fail:

```
    @Test
    fun testNodeInWindow() {
        rule.setContent {
            Window(
                onCloseRequest = {},
            ) {
                Text(
                    text = "Text",
                    modifier = Modifier.testTag("tag")
                )
            }
        }

        rule.onNodeWithTag("tag").assertExists()
    }
```

## Proposed Changes

  - When a `Window` or `DialogWindow` is created, it will add its `ComposeScene` as a child scene to its parent.
  - `ComposeScene.roots` will return its own `owners` and all of its descendent's `owners`.
 

## Testing

Test: Added unit tests to verify that nodes inside `Window` and `DialogWindow` can be found.

### CODE REVIEW

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt [hunk 1]
```diff
@@ -37,9 +37,7 @@ import org.jetbrains.skia.Canvas
 import org.jetbrains.skiko.SkiaLayer
 import org.jetbrains.skiko.currentNanoTime
 
-internal val LocalComposeScene = staticCompositionLocalOf<ComposeScene> {
-    error("CompositionLocal LocalComposeScene not provided")
```
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:41]: It was better than `!!`. What's the reason for this change?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt [hunk 2]
```diff
@@ -108,6 +106,16 @@ class ComposeScene internal constructor(
         invalidate,
     )
 
+    private val childScenes = mutableSetOf<ComposeScene>()
```
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: It should be cleaned in `close`

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt [hunk 3]
```diff
@@ -108,6 +106,16 @@ class ComposeScene internal constructor(
...
+
+    fun addChildScene(scene: ComposeScene) {
+        childScenes.add(scene)
```
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]:
```suggestion
        check(!isClosed) { "ComposeScene is closed" }
        childScenes.add(scene)
```

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 2 above)
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: Not sure about naming here.
1. child -> children
2. "child" implies ownership, but it's not

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt [hunk 4]
```diff
@@ -397,16 +398,21 @@ fun Window(
...
             create().apply {
+                parentScene?.addChildScene(scene)
                 this.compositionLocalContext = compositionLocalContext
                 this.exceptionHandler = windowExceptionHandlerFactory.exceptionHandler(this)
                 setContent(onPreviewKeyEvent, onKeyEvent, content)
             }
         },
-        dispose = dispose,
+        dispose = {
+            parentScene?.removeChildScene(it.scene)
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt:413]: I don't really like that one of important "integrity" related clean function is outside of the class

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 1 above)
Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt:41]: Because we're now accessing `LocalComposeScene` in a place where it hasn't been provided yet (at the root window/dialog). See discussion in Slack, too.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 2 above)
Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: When `close()` disposes the composition, the disposal code in `DialogWindow` and `Window` will be called, and it will remove the child.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 3 above)
Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: I don't see a case where this could happen, but sure, I'll add it just in case.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 2 above)
Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: Not sure why `childScenes` implies ownership ;-)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt (hunk 4 above)
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt:413]: Not sure what you mean. It's `Window`s responsibility to add it, so it also removes it.

**Comment by m-sasha** [comment]
Hmm, maybe the tests should be in `WindowTest` and `DialogTest` instead. What do you think?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt (hunk 4 above)
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt:413]: `Window` wasn't completely clean before :). It adds a native window to the native window list, and adds a new composition node to the SlotTable.

Registering a scene in the parent scene is similar to these.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt [hunk 5]
```diff
@@ -175,12 +173,43 @@ class ComposeScene internal constructor(
         }
     }
 
+    private val childScenes = mutableSetOf<ComposeScene>()
+
+    /**
+     * Adds a child [ComposeScene], so that nodes in it can be found in tests.
+     */
+    fun addChildScene(scene: ComposeScene) {
```
Comment (igordmn) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: This function doesn't have a strict defined behavior for external users.

Solutions:
- make it `internal` for now.
- rename to addRootForTest (but I am very reluctant to do that, because we still need to think about how this function can be useful for external user)
- make a parallel tree of `roots` that doesn't live in `ComposeScene` (it is very vague idea, I don't know how it should look yet)

The easiest option is the first option.

Comment (igordmn) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: We also don't have public API for adding new owners directly to the ComposeScene, because the API wasn't well thought out. The only current way to add an owner - via popup.

There were a thought - adding a new owner via a public function `scene.addRoot(content)`, which name is similar to `addChildScene`

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt (hunk 4 above)
Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/Window.desktop.kt:413]: I meant that it controls internal state of `parentScene`, but scene itself knows about creation and disposing events

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt (hunk 5 above)
Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt]: I think I'll make it `internal`.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/DesktopPopup.desktop.kt [hunk 6]
```diff
@@ -37,7 +37,7 @@ import kotlin.math.roundToInt
  */
 @Composable
 private fun rememberCursorPosition(): Offset? {
-    val scene = LocalComposeScene.current
+    val scene = LocalComposeScene.current!!
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/DesktopPopup.desktop.kt]: Can we check it with some meaningful error message? Just in case

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/PopupLayout.skiko.kt [hunk 7]
```diff
@@ -41,7 +41,7 @@ internal fun PopupLayout(
     onKeyEvent: ((KeyEvent) -> Boolean) = { false },
     content: @Composable () -> Unit
 ) {
-    val scene = LocalComposeScene.current
+    val scene = LocalComposeScene.current!!
```
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/PopupLayout.skiko.kt]: Same here

**Review by MatkovIvan — APPROVED** [review]
